### PR TITLE
KFSPTS-17867 Implement retries for failed ACH Extract rows

### PR DESCRIPTION
**Pull request by db483**

### CODE REVIEW

**Review by ajd299 — CHANGES_REQUESTED** [review]
Please see my feedback, note that many of the changes suggested are optional.

File: src/main/java/edu/cornell/kfs/pdp/batch/PayeeACHAccountExtractReportData.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+public class PayeeACHAccountExtractReportData {
+    
+    private List<PayeeACHAccountExtractFileResult> achAccountExtractFileResults;
+    private PayeeACHAccountExtractRetryResult achAccountExtractRetryResults;
+    
+    public PayeeACHAccountExtractReportData() {
+        achAccountExtractFileResults = new ArrayList<PayeeACHAccountExtractFileResult>();
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/pdp/batch/PayeeACHAccountExtractReportData.java:12]: updated

File: src/main/java/edu/cornell/kfs/pdp/batch/PayeeACHAccountExtractResult.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    private List<String> errors;
+    private Map<PayeeACHAccountExtractDetail, List<String>> errorEntries;
+    private List<PayeeACHAccountExtractDetail> successEntries;
+    private int numberOfRowsProcessedSuccessfully;
+    private int numberOfRowsWithFailures;
+    
+    public PayeeACHAccountExtractResult(){
+        errors = new ArrayList<String>();
+        errorEntries = new HashMap<PayeeACHAccountExtractDetail, List<String>>();
+        successEntries = new ArrayList<PayeeACHAccountExtractDetail>();
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/pdp/batch/PayeeACHAccountExtractResult.java:21]: updated

File: src/main/java/edu/cornell/kfs/pdp/batch/PayeeACHAccountExtractRetryResult.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package edu.cornell.kfs.pdp.batch;
+
+public class PayeeACHAccountExtractRetryResult extends PayeeACHAccountExtractResult{
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/pdp/batch/PayeeACHAccountExtractRetryResult.java:3]: updated

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractReportServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    public static String getFileNameWithoutPath(String fileName) {
+        String result = fileName;
+        if (StringUtils.contains(result, CUKFSConstants.SLASH)) {
+            result = StringUtils.substringAfterLast(fileName, CUKFSConstants.SLASH);
+        }
+        if (StringUtils.contains(result, CUKFSConstants.BACKSLASH)) {
+            result = StringUtils.substringAfterLast(result, CUKFSConstants.BACKSLASH);
+        }
+        return result;
+    }
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractReportServiceImpl.java:82]: I moved the code to CuBatchFileUtils. I created https://jira.cornell.edu/browse/KFSPTS-20973 to update RassReportServiceImpl.java

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractReportServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        reportWriterService.writeNewLines(1);
+        reportWriterService.writeNewLines(1);
+    }
+
+    private void writeErrorDetails(PayeeACHAccountExtractResult processingResult) {
+        reportWriterService.writeSubTitle("*** Entries processed with errors ***");
+        if (processingResult.getErrorEntries().isEmpty()) {
+            reportWriterService.writeFormattedMessageLine("* No entries processed with errors.");
+        } else {
+            for (PayeeACHAccountExtractDetail achDetail : processingResult.getErrorEntries().keySet().stream().sorted((o1,o2)-> o1.getId().compareTo(o2.getId())).collect(Collectors.toList())) {
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractReportServiceImpl.java:143]: updated

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java [hunk 6]
```diff
@@ -74,32 +81,52 @@
...
         LOG.info("processACHBatchDetails: Beginning processing of ACH input files");
+        loadACHBatchDetailFiles(achAccountExtractReportData);
         
+        payeeACHAccountExtractReportService.writeBatchJobReports(achAccountExtractReportData);
+
+        // For now, return true even if files or rows did not load successfully. Functionals will address the failed rows/files accordingly.
+        return true;
+    }
+    
+    private void loadACHBatchDetailFiles(PayeeACHAccountExtractReportData achAccountExtractReportData){
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java:109]: updated

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java [hunk 7]
```diff
@@ -201,12 +231,119 @@ public boolean processACHBatchDetails() {
...
+        LOG.info("loadACHBatchDetailPersisted: Retries that failed: " + numRetryFail);      
+    }
+    
+    private void addSuccessToReport(PayeeACHAccountExtractRetryResult processingRetryResult, PayeeACHAccountExtractDetail achDetail) {
+        processingRetryResult.addSuccessRow();
+        processingRetryResult.getSuccessEntries().add(achDetail);
+    }
+    
+    private void addErrorToReport(PayeeACHAccountExtractRetryResult processingRetryResult, PayeeACHAccountExtractDetail achDetail, String error) {
+        List<String> errors = new ArrayList<String>();
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java:291]: updated

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java [hunk 8]
```diff
@@ -201,12 +231,119 @@ public boolean processACHBatchDetails() {
...
+        List<PayeeACHAccountExtractDetail> persistedPayeeACHAccountExtractDetails = getPersistedPayeeACHAccountExtractDetails();
+        if (!persistedPayeeACHAccountExtractDetails.isEmpty()) {
+            return getSortedACHDetailsEligibleForRetry(persistedPayeeACHAccountExtractDetails);
+        }
+        return persistedPayeeACHAccountExtractDetails;
+    }
+
+    private List<PayeeACHAccountExtractDetail> getSortedACHDetailsEligibleForRetry(List<PayeeACHAccountExtractDetail> persistedPayeeACHAccountExtractDetails) {
+        List<PayeeACHAccountExtractDetail> achDetailsEligibleForRetry = new ArrayList<PayeeACHAccountExtractDetail>();
+        List<PayeeACHAccountExtractDetail> sortedAchDetailsEligibleForRetry = new ArrayList<PayeeACHAccountExtractDetail>();
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java:311]: updated

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java [hunk 9]
```diff
@@ -201,12 +231,119 @@ public boolean processACHBatchDetails() {
...
+        achDetailsEligibleForRetry = persistedPayeeACHAccountExtractDetails.stream().filter(a -> a.getRetryCount() < getMaxRetryCount())
+                .collect(Collectors.toList());
+        sortedAchDetailsEligibleForRetry = achDetailsEligibleForRetry.stream().sorted((o1, o2) -> o1.getId().compareTo(o2.getId()))
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java:314]: updated

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java [hunk 10]
```diff
@@ -201,12 +231,119 @@ public boolean processACHBatchDetails() {
...
+                .collect(Collectors.toList());
+        return sortedAchDetailsEligibleForRetry;
+    }
+
+    private int getMaxRetryCount() {
+        return Integer.valueOf(parameterService.getParameterValueAsString(PayeeACHAccountExtractStep.class, CUPdpParameterConstants.MAX_ACH_ACCT_EXTRACT_RETRY));
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java:320]: updated

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java [hunk 11]
```diff
@@ -201,12 +231,119 @@ public boolean processACHBatchDetails() {
...
+    }
+    
+    private void storeACHAccountExtractDetailForRetry(PayeeACHAccountExtractDetail achDetail, String status) {
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java:323]: updated

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java [hunk 12]
```diff
@@ -201,12 +231,119 @@ public boolean processACHBatchDetails() {
...
+        achDetail.setCreateDate(dateTimeService.getCurrentSqlDate());
+        achDetail.setStatus( CUPdpConstants.PayeeAchAccountExtractStatuses.OPEN);
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java:325]: removed

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java [hunk 13]
```diff
@@ -201,12 +231,119 @@ public boolean processACHBatchDetails() {
...
+    }
+
+    private void updateACHAccountExtractDetailRetryCount(PayeeACHAccountExtractDetail achDetail, int retryCount) {
+        achDetail.setLastUpdatedTimestamp(dateTimeService.getCurrentTimestamp());
+        achDetail.setRetryCount(retryCount);
+        businessObjectService.save(achDetail);
+    }
+
+    protected List<PayeeACHAccountExtractDetail> getPersistedPayeeACHAccountExtractDetails() {
+        List<PayeeACHAccountExtractDetail> persistedPayeeACHAccountExtractDetails = new ArrayList<PayeeACHAccountExtractDetail>();
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java:341]: updated

File: src/test/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImplTest.java [hunk 14]
```diff
@@ -541,6 +548,13 @@ private BusinessObjectService createMockBusinessObjectService() throws Exception
         
         return businessObjectService;
     }
+    
+    private DateTimeService createMockDateTimeService() throws Exception {
```
[the remark this answers is not in the document]
Comment (db483) [reply src/test/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImplTest.java:552]: removed

File: src/test/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImplTest.java [hunk 15]
```diff
@@ -570,6 +584,11 @@ private ConfigurationService createMockConfigurationService() throws Exception {
         
         return configurationService;
     }
+    
+
+    private PayeeACHAccountExtractReportService createMockPayeeACHAccountExtractReportService() throws Exception {
```
[the remark this answers is not in the document]
Comment (db483) [reply src/test/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImplTest.java:589]: removed

**Comment by db483** [comment]
I've added the requested changes but NOT ready yet.

**Comment by db483** [comment]
This is ready for review.

**Review by ajd299 — CHANGES_REQUESTED** [review]
One more small suggestion

File: src/main/java/edu/cornell/kfs/pdp/batch/PayeeACHAccountExtractResult.java [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    
+    private List<String> errors;
+    private Map<PayeeACHAccountExtractDetail, List<String>> errorEntries;
+    private List<PayeeACHAccountExtractDetail> successEntries;
+    private int numberOfRowsProcessedSuccessfully;
+    private int numberOfRowsWithFailures;
+    
+    public PayeeACHAccountExtractResult(){
+        errors = new ArrayList<String>();
+        errorEntries = new HashMap<PayeeACHAccountExtractDetail, List<String>>();
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/pdp/batch/PayeeACHAccountExtractResult.java:20]: updated

**Review by ajd299 — APPROVED** [review]
Code looks good, merging

**Pull request merged**